ESQL: CATEGORIZE as a BlockHash

nik9000 opened the pull request:
Re-implement `CATEGORIZE` in a way that works for multi-node clusters.

This requires that data is first categorized on each data node in a first pass, then the categorizers from each data node are merged on the coordinator node and previously categorized rows are re-categorized.

BlockHashes, used in HashAggregations, already work in a very similar way. E.g. for queries like `... | STATS ... BY field1, field2` they map values for `field1` and `field2` to unique integer ids that are then passed to the actual aggregate functions to identify which "bucket" a row belongs to. When passed from the data nodes to the coordinator, the BlockHashes are also merged to obtain unique ids for every value in `field1, field2` that is seen on the coordinator (not only on the local data nodes).

Therefore, we re-implement `CATEGORIZE` as a special BlockHash.

To choose the correct BlockHash when a query plan is mapped to physical operations, the `AggregateExec` query plan node needs to know that we will be categorizing the field `message` in a query containing `... | STATS ... BY c = CATEGORIZE(message)`. For this reason, _we do not extract the expression_ `c = CATEGORIZE(message)` into an `EVAL` node, in contrast to e.g. `STATS ... BY b = BUCKET(field, 10)`. The expression `c = CATEGORIZE(message)` simply remains inside the `AggregateExec`'s groupings.

**Important limitation:** For now, to use `CATEGORIZE` in a `STATS` command, there can be only 1 grouping (the `CATEGORIZE`) overall.

nik9000 committed:
WIP

nik9000 committed:
Merge branch 'main' into esql_blockh2

nik9000 committed:
Stay plz:

nik9000 committed:
Update docs/changelog/114317.yaml

elasticsearchmachine commented:
Hi @nik9000, I've created a changelog YAML for you.

alex-spies committed:
Make things compile

somehow

alex-spies committed:
Move new block hashes to typical location

This makes them easier to be tested.

alex-spies committed:
Add almost passing test

jan-elastic committed:
Close stuff (makes unit test pass)

jan-elastic committed:
undo making BytesRefBlockHash public

jan-elastic committed:
Move Categorize BlockHash tests to separate file

jan-elastic committed:
Unit test for CategorizedIntermediateBlockHash.

jan-elastic committed:
Unit test with Driver

jan-elastic committed:
Add aggregator to unit test

jan-elastic committed:
Passing unit test with driver and aggregators

jan-elastic committed:
Improve code

jan-elastic committed:
Fix ES compile errors

costin commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/ElementType.java:
```diff
@@ -24,6 +24,8 @@ public enum ElementType {
     NULL("Null", (blockFactory, estimatedSize) -> new ConstantNullBlock.Builder(blockFactory)),
 
     BYTES_REF("BytesRef", BlockFactory::newBytesRefBlockBuilder),
+    CATEGORY_RAW("CategoryRaw", BlockFactory::newBytesRefBlockBuilder),
+    CATEGORY_INTERMEDIATE("CategoryIntermediate", BlockFactory::newBytesRefBlockBuilder),
```
We talked in the past about associating an encoder/decoder to the byterefs blocks to encapsulate the specific content without introducing new types.
Would this work here?

nik9000 replied:
I *think* we can get away without introducing a new type by being more explicit when we're in the building code and building the `BlockHash` implementation "by hand" rather than going through the type generic methods. After all, this method is super special and those generic methods are built to handle the data regardless of it's source.

alex-spies replied:
To wrap this up: The current iteration doesn't put the information into the bytesref blocks - instead, we just don't extract the `CATEGORIZE` function into an eval and leave it inside the `Aggregate` logical plan. Later down the road, when mapping to physical operations, we detect the `CATEGORIZE` in the `AggregateExec` and use a custom block hash for `CATEGORIZE.

jan-elastic committed:
Update Categorize function/evaluator (works end2end in discover)

jan-elastic committed:
fix CategorizeTests

jan-elastic committed:
Fix

jan-elastic committed:
Misc fixes to make CSV test work

alex-spies committed:
Release blocks in CategorizedIntermediateBlockHash

jan-elastic committed:
Decode category

alex-spies commented on x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/type/DataType.java:
```diff
@@ -192,7 +192,10 @@ public enum DataType {
-    PARTIAL_AGG(builder().esType("partial_agg").unknownSize());
+    PARTIAL_AGG(builder().esType("partial_agg").unknownSize()),
+
+    // NOCOMMIT: if we keep this DataType, what should esType be?
+    CATEGORY(builder().esType("category").unknownSize());
```
++ Let's' see if we can avoid introducing a new data type. Adding a new type requires touching lots of things to make all work consistently.

alex-spies commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/AbstractCategorizeBlockHash.java:
```diff
@@ -0,0 +1,98 @@
+public abstract class AbstractCategorizeBlockHash extends BlockHash {
+    // TODO: this should probably also take an emitBatchSize
+    private final int channel;
+    private final boolean outputPartial;
```
nit: I think `outputPartial` may not be needed as we have separate subclasses for the initial and final versions of the block hash.

jan-elastic replied:
Disclaimer: I don't know the ES|QL aggs in details, but:

- `AggregatorMode.SINGLE` suggests we need a raw hash with final output; and
- `AggregatorMode.INTERMEDIATE` suggests we need an intermediate hash with intermediate output

I'd guess the "merging aggregated data" can happen stages. So:

raw data -> raw hash -> intermediate hash -> intermediate hash -> final output

alex-spies replied:
Yep, you're correct - thanks, please disregard!

alex-spies commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/BlockHash.java:
```diff
@@ -127,14 +132,17 @@ public static BlockHash buildPackedValuesBlockHash(List<GroupSpec> groups, Block
      * Creates a specialized hash table that maps a {@link Block} of the given input element type to ids.
      */
-    private static BlockHash newForElementType(int channel, ElementType type, BlockFactory blockFactory) {
+    private static BlockHash newForElementType(int channel, ElementType type, AggregatorMode aggregatorMode, BlockFactory blockFactory) {
```
Hm, maybe a more natural way to pass in the info that we need a category block hash would be to alter `GroupSpec` and instead of `ElementType`, use a new class/enum that either is a group based on a simple data type or a categorization?

jan-elastic replied:
Where should the logic go that indicates that the result of the `Categorize` function is a `BytesRef` that should be interpreted as a `Category` and go to the `CategoryRawBlockHash`? 

And should the no-op `Categorize` function stay or go?

alex-spies replied:
I'll need to take a closer look, but I _think_ in `AbstractPhysicalOperationProviders.groupingPhysicalOperation` we should have all required information to choose the correct block hash. Currently, this info is passed on by passing the aggregator mode and the group spec (resp. its data type) to the `HashAggregationOperatorFactory`'s constructor, but we could pass this info on by different means as well - if I'm correct.

nik9000 replied:
I was thinking we'd avoid calling these functions entirely when we know we're dealing with CATEGORIZE. But I haven't found the spot to make that choice.

alex-spies commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/CategorizedIntermediateBlockHash.java:
```diff
@@ -0,0 +1,68 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class CategorizedIntermediateBlockHash extends AbstractCategorizeBlockHash {
```
Suggestion:
```suggestion
public class CategorizedFinalBlockHash extends AbstractCategorizeBlockHash {
```

In the future, there may be a proper intermediate version of this block hash because we may have 3 aggregation steps: initial on the level of shards, intermediate on the level of the data node, final on the coordinator node.

jan-elastic replied:
I think the proper intermediate version already exists. It's this `BlockHash` with `outputPartial=true`.

alex-spies replied:
You are right!

alex-spies reviewed (commented):
Alright, I think we're close!

Once we remove the workaround that adds a new data type and add a couple more tests, I think we'll be able to merge this.

nik9000 commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/AbstractCategorizeBlockHash.java:
```diff
@@ -0,0 +1,98 @@
+        return IntVector.range(0, categorizer.getCategoryCount(), blockFactory);
+    }
+
+    @Override
+    public BitArray seenGroupIds(BigArrays bigArrays) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public final ReleasableIterator<IntBlock> lookup(Page page, ByteSizeValue targetBlockSize) {
+        throw new UnsupportedOperationException();
+    }
```
Correct. We might want to fork the interface for `lookup` off of BlockHash. It's part of the infrastructure for the hash lookup. It's fine to skip this here.

At some point we should think about this thing in combination with INLINESTATS. Right now INLINESTATS forces the PackedValuesBlockHash which is slow but compatible with the kind of null handling that lookup expected. One day we'll tweak things and you can use any of these for INLINESTATS. But that's not today at all.

nik9000 commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/AbstractCategorizeBlockHash.java:
```diff
@@ -0,0 +1,98 @@
+            return blockFactory.newConstantNullBlock(categorizer.getCategoryCount());
+        }
+        try (BytesStreamOutput out = new BytesStreamOutput()) {
+            // TODO be more careful here.
```
I totally hacked this thing together. I haven't a clue if it's properly correct. I mean, it works, but it's worth another set of eyes on it.

jan-elastic replied:
I wasn't sure what you meant with "be more careful here" but it lgtm

nik9000 commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/BlockHash.java:
```diff
@@ -127,14 +132,17 @@ public static BlockHash buildPackedValuesBlockHash(List<GroupSpec> groups, Block
             case LONG -> new LongBlockHash(channel, blockFactory);
             case DOUBLE -> new DoubleBlockHash(channel, blockFactory);
             case BYTES_REF -> new BytesRefBlockHash(channel, blockFactory);
+            case CATEGORY -> aggregatorMode.isInputPartial()
```
I'm hoping we can avoid passing this in at all.

nik9000 commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/HashAggregationOperator.java:
```diff
@@ -39,14 +40,15 @@ public class HashAggregationOperator implements Operator {
             return new HashAggregationOperator(
                 aggregators,
-                () -> BlockHash.build(groups, driverContext.blockFactory(), maxPageSize, false),
+                () -> BlockHash.build(groups, aggregatorMode, driverContext.blockFactory(), maxPageSize, false),
```
Yeah - I'd prefer to change this call *somehow*.

ivancea replied:
I wonder if we should make this change in the follow-up

nik9000 commented on x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/aggregation/blockhash/BlockHashTestCase.java:
```diff
@@ -0,0 +1,34 @@
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class BlockHashTestCase extends ESTestCase {
```
Probably worth some javadoc. And maybe some abstract methods to force us to test some BlockHash stuff in common? I dunno. We have a lot of that from the randomized test.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/grouping/Categorize.java:
```diff
@@ -29,17 +21,17 @@
 import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.isString;
 
+// NOCOMMIT
+// This Categorize function is a no-op, passing on BytesRefs and rebranding them as categories.
+// I doubt that's the way to go, but it gets the job done...
```
I think it should probably throw an exception from `toEvaluator` - like we avoid building it at all. Or it can noop, that's fine too if that's easier. But it really gets "shifted" into the agg.

nik9000 commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/grouping/CategorizeTests.java:
I imagine some of the test cases for type handling are still ok. But the evaluation cases aren't.

nik9000 replied:
I'm not super sure what to do with that.

alex-spies committed:
Add test case with expression inside CATEGORIZE

alex-spies committed:
Disallow evaluating CATEGORIZE outside aggs

alex-spies commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/categorize.csv-spec:
Another interesting edge case: `CATEGORIZE(message)` where `message is a `text` field instead of `keyword`.

nik9000 replied:
> different kinds of aggregations, like `SUM`, `TOP`, `COUNT_DISTINCT` used with `BY CATEGORIZE(message)`

Mostly this is just paranoia. We're just grabbing a smattering of different aggs to make sure we don't break something. `AVG` is quite useful because of it's funny rewrites here. 

I can grab a few of these.

nik9000 replied:
I fixed multivalue handling and pushed a test for it. This ones doesn't work because of the new type:
```
cases where we do something with the categories, like stats ... by c = categorize(...) | eval y = to_upper(c)
```
Just removing the new type *should* fix it.

alex-spies replied:
Thanks @nik9000 for grabbing these, that's awesome!

I have two more suggestions:
- `STATS BY CATEGORIZE(message)` (no agg function, only categorization)
- `STATS by a = c, b = c BY c = CATEGORIZE(message)`

alex-spies replied:
And another from @ivancea 's positive verifier tests:
`from test | STATS MV_COUNT(cat), COUNT(*) BY cat = CATEGORIZE(first_name)`

alex-spies replied:
- [ ] Similar, but without aliasing:
`| stats mv_count(bucket(x, 1)) by bucket(x, 1)`

alex-spies replied:
Another one:

- [ ] `| STATS BY c = CATEGORIZE(message) | EVAL cc = TO_UPPER(c)`

nik9000 replied:
It looks like we didn't get to quite a few of these. More fodder for a follow-up issue.

alex-spies replied:
I summarized the remainder [here](https://github.com/elastic/elasticsearch/pull/114317#pullrequestreview-2464510445), and we'll defintely have a follow-up!

alex-spies committed:
WIP: wire in CATEGORIZE without new datatype

alex-spies committed:
Wire in CATEGORIZE without needing new datatype

Make it so the information that we need a to use a categorizing block
hash is local to the AbstractPhysicalOperationProviders. This can now be
conveyed by putting this - somehow - into the GroupSpec class, and then
the Category data type shouldn't be required anymore.

nik9000 committed:
multivalued fields

nik9000 committed:
ROW MV

nik9000 committed:
More cases

nik9000 committed:
unicode

nik9000 committed:
More cases

nik9000 committed:
Drop all columns

jan-elastic commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/AbstractCategorizeBlockHash.java:
```diff
@@ -0,0 +1,98 @@
+
+import java.io.IOException;
+
+public abstract class AbstractCategorizeBlockHash extends BlockHash {
```
I'm wondering whether we should replace `AbstractCategorizeBlockHash` / `CategorizeRawBlockHash` / `CategorizeIntermediateBlockHash` by just one `CategorizeBlockHash` with a constructor param `inputPartial` (next to the existing `outputPartial`)?

@nik9000 @alex-spies WDYT?

alex-spies replied:
Seems reasonable!

jan-elastic commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/categorize.csv-spec:
```diff
@@ -1,14 +1,228 @@
+           1 | .*?Disconnected.+?banana.*?
+;
+
+on CONCAT
```
duplicate test

jan-elastic commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/categorize.csv-spec:
```diff
@@ -1,14 +1,228 @@
+
+on REVERSE(CONCAT())
+required_capability: categorize
+# NOCOMIT needs fn_reverse_grapheme
```
NOCOMIT -> NOCOMMIT 😃

nik9000 replied:
disaster

ivancea replied:
It works, removing the commit. I could remove the test, as it's indeed quite redundant, but it's there, so whatever (?)

nik9000 committed:
ON

ivancea committed:
Move BlockHash creation to Grouping function

ivancea committed:
Remove CATEGORY DataType and ElementType

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/AbstractPhysicalOperationProviders.java:
```diff
@@ -138,7 +140,12 @@ else if (aggregatorMode == AggregatorMode.INITIAL || aggregatorMode == Aggregato
+
+                var groupSpec = Alias.unwrap(group) instanceof GroupingFunction groupingFunction
+                    ? new GroupSpec(groupInput == null ? null : groupInput.channel(), groupAttribute, groupingFunction)
+                    : new GroupSpec(groupInput == null ? null : groupInput.channel(), groupAttribute, null);
```
Maybe we can also just expand the definition of `GroupSpec` to not rely on an attribute:

From
```
private record GroupSpec(Integer channel, Attribute attribute)
```
to
```
private record GroupSpec(Integer channel, NamedExpression groupExpression)
```

Or, if that's too annoying to wire in, we can also just add a boolean here to indicate whether or not we categorize on this group based on whether or not the grouping function is `instanceOf Categorize`.

ivancea replied:
The attribute is being used in other place (`ordinalGroupingOperatorFactory`). So I left both for now: `Integer, Attribute, Expression (Instead of GroupingFunction)`.
No strong opinions here, so I would leave it that way for now.

About storing a boolean instead of the expression, I feel like the expression is as simple, and is not "too specific" for categorization

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/AbstractPhysicalOperationProviders.java:
```diff
@@ -172,6 +179,13 @@ else if (aggregatorMode == AggregatorMode.INITIAL || aggregatorMode == Aggregato
                     context
                 );
             } else {
+                if (groupSpecs.stream().anyMatch(g -> g.groupingFunction != null)) {
+                    // TODO: DONT DO THIS, bucket should work: assert groupSpecs.size() == 1 : "Only one grouping function is supported";
+                    // Maybe ignore Bucket directly with instanceof
+                    // Maybe add some method to GroupingFunction to get the hash?
+                    // Maybe a method returning a Supplier of HashBlock?
+                }
```
Just adding a reminder here because that's an important `NOCOMMIT`.

ivancea replied:
Removed! This was an initial idea I discarded, and forgot to delete it

alex-spies reviewed (commented):
Thanks @ivancea , progress! We no longer rely on the `Category` data type and can get rid of that :)

My only qualm is that having the method `Categorize.toBlockHash` makes our abstractions a bit leaky and I'd like to avoid that. I left a suggestion for an alternative which we can get back to before merging.

alex-spies reviewed (commented):
Oh, and there we go, `Category` datatype gone, thanks @ivancea ! :rocket:

ivancea committed:
Format all

ivancea committed:
Removed unused code and added new test using CATEGORY() return value in EVAL

ivancea committed:
Remove ToBlockHash interface, and simplify logic for Categorize

ivancea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/AbstractPhysicalOperationProviders.java:
```diff
@@ -299,12 +313,17 @@ private void aggregatesToFactory(
+            return new BlockHash.GroupSpec(
+                channel,
+                elementType(),
+                groupingFunction instanceof ToBlockHash toBlockHash ? toBlockHash : null
```
I just removed it. I tried to look ahead with this interface, but it's true that we don't have enough info yet, and it didn't fit too well.

It's a bit more specific now, but it's quite simpler. At least until we refactor it for the multi-group feature

ivancea committed:
Merge branch 'main' into esql_blockh2

# Conflicts:
#	x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/BlockHash.java
#	x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/aggregation/blockhash/BlockHashTests.java
#	x-pack/plugin/esql/qa/testFixtures/src/main/java/org/elasticsearch/xpack/esql/CsvAssert.java
#	x-pack/plugin/esql/src/main/generated/org/elasticsearch/xpack/esql/expression/function/grouping/CategorizeEvaluator.java
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/CombineProjections.java
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/AbstractPhysicalOperationProviders.java

ivancea committed:
Remove missing CATEGORY enum constant from Csv tests

ivancea committed:
Keep alias on ReplaceAggNested Categorize grouping

ivancea committed:
Do not null/propagate fold Categorize

ivancea committed:
Fixed Categorize layout using source field, and added failing tests with aliases

ivancea committed:
Added failing alias CSV tests for Categorize

ivancea committed:
Make CombineProjections work with expressions in the grouping, like Categorize

ivancea committed:
Added Categorize tests for FoldNull and nested aggs rules

ivancea committed:
Format with spotlessApply

ivancea committed:
Fix benchmark compilation

ivancea committed:
Fix AttributeSet add() return value

ivancea committed:
Fix/Remove NOCOMMITs

ivancea committed:
Fix Javadoc HTML p tag not allowing nested blocks

ivancea committed:
Fixed tests compilation and added more csv tests for other commands

ivancea committed:
Remove remaining NOCOMMITs

ivancea committed:
Format

ivancea committed:
Updated Categorize tests to not fail on evaluator creation

ivancea commented on x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/expression/AttributeSet.java:
```diff
@@ -113,7 +113,7 @@ public <T> T[] toArray(T[] a) {
 
     @Override
     public boolean add(Attribute e) {
-        return delegate.put(e, PRESENT) != null;
+        return delegate.put(e, PRESENT) == null;
     }
```
Changed this to follow the Set conventions. It was probably a typo.

We should change it in the QL counterpart too. I haven't done it yet, as to avoid mixing more things here

nik9000 replied:
I might pull this one character change out into it's own PR.....

nik9000 replied:
Let me be clear - I'd probably pull this into it's own PR because there are like 1232134 callers of this method. Do they use the return value? That's hard to tell because this is a collection implementation. If you land it in one PR, even as a one line change. Or a one line change with a single test, then we will have an easier time tracking tricky changes to it.

I'm just imagining `git bisect` pointing to a PR that does *just* this thing vs a PR that does, well, all of CATEGORIZE.

ivancea replied:
https://github.com/elastic/elasticsearch/pull/117367

ivancea replied:
PS: Absolutely, I'm with you here, it makes all the sense!

ivancea committed:
Merge branch 'main' into esql_blockh2

alex-spies reviewed (commented):
Heya, this is beginning to look mighty good.

I need to thoroughly review the optimizer rule changes as these could affect already working queries. But the stuff put in place to make CATEGORIZE work looks nice and we should be able to un-draft/merge this soon.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/CombineProjections.java:
```diff
@@ -137,23 +141,33 @@ private static List<NamedExpression> combineProjections(List<? extends NamedExpr
+        List<? extends NamedExpression> upperGroupings,
         List<? extends NamedExpression> lowerProjections
     ) {
         // Collect the alias map for resolving the source (f1 = 1, f2 = f1, etc..)
-        AttributeMap<Attribute> aliases = new AttributeMap<>();
+        AttributeMap<Expression> aliases = new AttributeMap<>();
         for (NamedExpression ne : lowerProjections) {
-            // Projections are just aliases for attributes, so casting is safe.
-            aliases.put(ne.toAttribute(), (Attribute) Alias.unwrap(ne));
+            // record the alias
+            aliases.put(ne.toAttribute(), Alias.unwrap(ne));
         }
```
nit: I do not think this change is necessary. The lower projections should still only be aliases that point to attributes - that's all a projection is allowed to contain (even though we don't make this invariant explicit).

nik9000 replied:
@ivancea Alex had a few comments on this file. Did y'all finish these? Or should these also go to a follow-up?

alex-spies replied:
Should be addressed in https://github.com/elastic/elasticsearch/pull/117882

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/CombineProjections.java:
```diff
@@ -137,23 +141,33 @@ private static List<NamedExpression> combineProjections(List<? extends NamedExpr
+        AttributeSet seen = new AttributeSet();
+        List<Expression> replaced = new ArrayList<>();
+        for (NamedExpression ne : upperGroupings) {
+            // Duplicated attributes are ignored.
+            if (ne instanceof Attribute attribute) {
+                var newExpression = aliases.resolve(attribute, attribute);
+                if (newExpression instanceof Attribute newAttribute && seen.add(newAttribute) == false) {
+                    // Already seen, skip
+                    continue;
+                }
+                replaced.add(newExpression);
+            } else {
+                // For grouping functions, this will replace nested properties too
+                replaced.add(ne.transformUp(Attribute.class, a -> aliases.resolve(a, a)));
+            }
         }
-        return new ArrayList<>(replaced);
+        return replaced;
```
nit: I think this change can become a bit smaller with less special-casing - which I find to be easier to reason about and verify correctness.

```
    private static List<Expression> combineUpperGroupingsAndLowerProjections(
        List<? extends NamedExpression> upperGroupings,
        List<? extends NamedExpression> lowerProjections
    ) {
        // Collect the alias map for resolving the source (f1 = 1, f2 = f1, etc..)
        AttributeMap<Attribute> aliases = new AttributeMap<>();
        for (NamedExpression ne : lowerProjections) {
            // record the alias
            aliases.put(ne.toAttribute(), (Attribute) Alias.unwrap(ne));
        }

        // Replace any matching attribute directly with the aliased attribute from the projection.
        AttributeMap<NamedExpression> replaced = new AttributeMap<>();
        for (NamedExpression ne : upperGroupings) {
            // All substitutions happen before; groupings must be attributes at this point
            // except for CATEGORIZE which will be an alias like `c = CATEGORIZE(attribute)`
            NamedExpression transformed = (NamedExpression) ne.transformUp(Attribute.class, a -> aliases.resolve(a, a));
            replaced.add(transformed.toAttribute(), transformed);
        }
        return new ArrayList<>(replaced.values());
    }
```

I ran the tests with my version and they seemed to pass the same :)

alex-spies reviewed (commented):
Heya, great job iterating on this @ivancea !

I need to do at least one last round, but I had a chance to look at the changes to optimization rules and can confirm that the changes should only affect code paths for queries that contain a `CATEGORIZE` somewhere - so this PR looks safe to be merged in general and I think we should un-draft it.

ivancea committed:
Fixed union_types CSV tests, failing after new dataset was addition

ivancea committed:
Merge branch 'main' into esql_blockh2

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:
```diff
@@ -1203,6 +1204,26 @@ public void testCombineProjectionWithAggregationFirstAndAliasedGroupingUsedInAgg
+
+        var limit = as(plan, Limit.class);
+        var agg = as(limit.child(), Aggregate.class);
+        assertThat(Expressions.names(agg.aggregates()), contains("s", "cat"));
+        assertThat(Expressions.names(agg.groupings()), contains("cat"));
```
I think the assertions here are insufficient because this would also pass if `CombineProjection` is not applied at all. IMO we should assert that
- before the `agg` is the `EsRelation`, and
- the `agg`'s groupings is not only named "cat", but also that it's an Alias that contains a Categorize that points to the FieldAttribute `first_name`.

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:
```diff
@@ -3909,6 +3930,37 @@ public void testNestedExpressionsInGroups() {
+    /**
+     * Expects
+     * Limit[1000[INTEGER]]
+     * \_Aggregate[[emp_no%2{r}#6],[COUNT(salary{f}#12) AS c, emp_no%2{r}#6]]
+     *   \_Eval[[emp_no{f}#7 % 2[INTEGER] AS emp_no%2]]
+     *     \_EsRelation[test][_meta_field{f}#13, emp_no{f}#7, first_name{f}#8, ge..]
+     */
```
The expectation here is out of date for the current test.

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:
```diff
@@ -3909,6 +3930,37 @@ public void testNestedExpressionsInGroups() {
+        var evalFieldAlias = as(eval.fields().get(0), Alias.class);
+        var evalField = as(evalFieldAlias.child(), Concat.class);
+
+        assertThat(categorize.field(), is(evalFieldAlias.toAttribute()));
+        assertThat(evalField.source().text(), is("CONCAT(first_name, \"abc\")"));
+        assertThat(categorizeAlias.source(), is(evalFieldAlias.source()));
```
I'd also assert the name of `evalFieldAlias`. The name cannot be arbitrary due to potential shadowing. The fact that we use the same name as the name of the group is important for this not to happen.

alex-spies reviewed (commented):
Ok, I finished reviewing the optimizer changes now!

alex-spies commented on x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/aggregation/blockhash/CategorizeBlockHashTests.java:
```diff
@@ -0,0 +1,401 @@
+            page.releaseBlocks();
+        }
+
+        // TODO: randomize and try multiple pages.
+        // TODO: assert the state of the BlockHash after adding pages. Including the categorizer state.
+        // TODO: also test the lookup method and other stuff.
```
I think this is still an important thing to do before merging. Esp. randomizing the tests a bit and testing in cases of multiple pages that are fed to the block hashes.

Except the lookup method, this one is unsupported and that's fine - let's just remove the `TODO`.

More things worth testing:
- handling of nulls
- handling of multivalues

C.f. the standard BlockHashTests that have a lot of this stuff (although the test infrastructure may not apply super well to the Categorize block hashes).

nik9000 replied:
I think we still need these. Do we have a tracking issue for follow up so we don't lose these? It's really nice to have a little check list. Checking boxes is fun.

alex-spies replied:
Also summarized [here](https://github.com/elastic/elasticsearch/pull/114317#pullrequestreview-2464510445); multivalues and nulls are already addressed here: https://github.com/elastic/elasticsearch/pull/117655/files

alex-spies commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/AbstractCategorizeBlockHash.java:
```diff
@@ -0,0 +1,98 @@
+
+    @Override
+    public BitArray seenGroupIds(BigArrays bigArrays) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public final ReleasableIterator<IntBlock> lookup(Page page, ByteSizeValue targetBlockSize) {
+        throw new UnsupportedOperationException();
```
nit: let's give these some useful error messages if we ever end up here - and maybe an explanation as a comment _why_ it's fine that we do not support these.

ivancea replied:
@nik9000 You may have more context about this. Is this something we didn't "want" to implement _yet_, or something that can't be done with this HashBlock?

nik9000 replied:
This is powered hash joins. Originally when I built the infrastructure for hash joins I thought *any* `BlockHash` could power the joins via this lookup. But we've been spending a lot less time on hash join lately. Maybe we'll get back to it soon? Anyway, as it is, only PackedValuesBlockHash actually implements `lookup`. I think eventually we will make the `lookup` method work for the other `BlockHash` implementations. 

Except maybe the `CATEGORIZE` one. I don't know how it'd work or if it'd make sense. I think this method is confusing and it'd be nice to remove it from `BlockHash` for now and fix the compilation errors.

alex-spies commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/CategorizeRawBlockHash.java:
```diff
@@ -0,0 +1,129 @@
+                        result.appendInt(process(vBlock.getBytesRef(first, vScratch)));
+                        continue;
+                    }
+                    int end = first + count;
+                    for (int i = first; i < end; i++) {
+                        result.appendInt(process(vBlock.getBytesRef(i, vScratch)));
+                    }
```
Going by the general BlockHashTests, I think the fact that our ordinals that we return do not have the same multi-value counts as the input block seems wrong. I'm confused how this doesn't lead to problems.

alex-spies commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/topn/KeyExtractor.java:
```diff
@@ -24,9 +24,6 @@ interface KeyExtractor {
     int writeKey(BreakingBytesRefBuilder key, int position);
 
     static KeyExtractor extractorFor(ElementType elementType, TopNEncoder encoder, boolean ascending, byte nul, byte nonNul, Block block) {
-        if (false == (elementType == block.elementType() || ElementType.NULL == block.elementType())) {
-            throw new IllegalArgumentException("Expected [" + elementType + "] but was [" + block.elementType() + "]");
-        }
```
@ivancea this is important to double check IMHO.

ivancea replied:
Yep, I think this was removed because of the new DataType. I think we can restore it. On it!

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/AbstractPhysicalOperationProviders.java:
```diff
@@ -182,6 +191,7 @@ else if (aggregatorMode == AggregatorMode.INITIAL || aggregatorMode == Aggregato
      * It's similar to the code above (groupingPhysicalOperation) but ignores the factory creation.
      */
     public static List<Attribute> intermediateAttributes(List<? extends NamedExpression> aggregates, List<? extends Expression> groupings) {
+        // TODO: needs to take CATEGORIZE into account: the intermediate agg's output is BytesRef, unlike the final keyword output.
```
I changed my mind: this can be done in a follow-up, but let's not forget - going back to this to figure it out later will be annoying :/

alex-spies reviewed (commented):
Ok, all done now in terms of reviewing!

Thanks again @ivancea , this is looking good. I think this is close.

A couple of things should be addressed before merging - especially any remarks on tests for functionality this PR already covers.

But for anything else - especially edge cases that this PR does not yet solve, @ivancea feel free to aggressively put that into follow-up tasks as you see fit.

There are also still unaddressed review comments from previews rounds - just pointing this out here because these are easy to overlook due to being collapsed away :)

ivancea committed:
Mute eval const failing tests until issue is fixed

ivancea committed:
Restored pre-DataType-change pieces of code (Extractors, BlockHash AggregatorMode parameter)

ivancea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/AbstractPhysicalOperationProviders.java:
```diff
@@ -324,6 +342,7 @@ public abstract Operator.OperatorFactory ordinalGroupingOperatorFactory(
         PhysicalOperation source,
         AggregateExec aggregateExec,
         List<GroupingAggregator.Factory> aggregatorFactories,
+        AggregatorMode aggregatorMode,
```
The Test provider eventually uses `BlockHash.build()`, which now requires the mode. Weird, but I don't think it's worth it to try to remove it

ivancea committed:
Added javadocs and some tests with mixed cases (nulls, multivalue...)

ivancea commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/CategorizedIntermediateBlockHash.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    @Override
+    public void add(Page page, GroupingAggregatorFunction.AddInput addInput) {
```
Added some javadocs around the Categorize hashblocks

ivancea committed:
Some docs and refactors on CategorizedIntermediateHashBlock

ivancea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/PropagateEvalFoldables.java:
```diff
@@ -32,6 +33,10 @@ public LogicalPlan apply(LogicalPlan plan) {
         // collect aliases bottom-up
         plan.forEachExpressionUp(Alias.class, a -> {
             var c = a.child();
+            if (c instanceof Categorize) {
+                // Categorize cannot be folded
+                return;
+            }
```
Jesus Christ, this logic makes no sense, my bad.
I'll just make Categorize non-foldable, and add that comment there :hehe:

ivancea replied:
Just double checked: The condition isn't needed because this rule doesn't replace anything in aggregations. Only filters and evals. So the grouping remains untouched

ivancea committed:
Make Categorize non-foldable

ivancea committed:
Format

ivancea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/ReplaceAggregateNestedExpressionWithEval.java:
```diff
@@ -43,18 +46,43 @@ protected LogicalPlan rule(Aggregate aggregate) {
+                if (as.child() instanceof Categorize cat) {
+                    if (cat.field() instanceof Attribute == false) {
+                        groupingChanged = true;
+                        var catAs = new Alias(as.source(), as.name(), cat.field());
+                        var attr = catAs.toAttribute();
```
In this specific case, the variable name matches the type. So I don't see an issue with it 👀

ivancea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/ReplaceAggregateNestedExpressionWithEval.java:
```diff
@@ -43,18 +46,43 @@ protected LogicalPlan rule(Aggregate aggregate) {
+                        evalNames.put(catAs.name(), attr);
+                        Categorize replacement = cat.replaceChildren(List.of(attr));
+                        newGroupings.set(i, as.replaceChild(replacement));
+                        ReferenceAttribute ref = new ReferenceAttribute(
```
That catAs (Which I'll rename, as it's the Categorize field really) is "moved" here to the eval. So, is it really considered synthetic?

ivancea replied:
Thinking again about it, the name itself is new indeed. So yeah, makes sense. Changing it!

ivancea committed:
ReplaceNestedAgg ruie cleanup

ivancea committed:
Merge branch 'main' into esql_blockh2

# Conflicts:
#	muted-tests.yml

ivancea committed:
Added failing tests for empty strings throwing NPE

ivancea commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/categorize.csv-spec:
```diff
@@ -1,14 +1,432 @@
+           1 | .*?disconnected.*?
+;
+
+on const empty string
```
This is failing with an NPE:
```
java.lang.NullPointerException: Cannot invoke "org.elasticsearch.xpack.ml.aggs.categorization.TokenListCategory.getId()" because the return value of "org.elasticsearch.xpack.ml.aggs.categorization.TokenListCategorizer$CloseableTokenListCategorizer.computeCategory(String, org.elasticsearch.xpack.ml.job.categorization.CategorizationAnalyzer)" is null
	at org.elasticsearch.compute.aggregation.blockhash.CategorizeRawBlockHash$CategorizeEvaluator.process(CategorizeRawBlockHash.java:127)
	at org.elasticsearch.compute.aggregation.blockhash.CategorizeRawBlockHash$CategorizeEvaluator.eval(CategorizeRawBlockHash.java:120)
	at org.elasticsearch.compute.aggregation.blockhash.CategorizeRawBlockHash$CategorizeEvaluator.eval(CategorizeRawBlockHash.java:89)
	at org.elasticsearch.compute.aggregation.blockhash.CategorizeRawBlockHash.add(CategorizeRawBlockHash.java:54)
	at org.elasticsearch.compute.operator.HashAggregationOperator.addInput(HashAggregationOperator.java:169)
        ...
```

ivancea committed:
Rename sample_data_mv

ivancea committed:
Ignore failing empty string tests

ivancea committed:
Undo union-types tests changes

ivancea committed:
Format

ivancea commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/sample_data_mv.csv:
```diff
@@ -0,0 +1,8 @@
+2023-10-23T13:53:55.832Z,172.21.3.15 ,5033755,[Connection error, Banana]
+2023-10-23T13:52:55.015Z,172.21.3.15 ,8268153,[Connection error, Banana]
+2023-10-23T13:51:54.732Z,172.21.3.15 , 725448,[Connection error, Banana]
+2023-10-23T13:33:34.937Z,172.21.0.5  ,1232382,[Disconnected, Banana]
```
Added a case using employees index, which has a mix of all

ivancea commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/union_types.csv-spec:
```diff
@@ -936,6 +936,13 @@ sample_data          | 2023-10-23T13:51:54.732Z  |  172.21.3.15   |  725448
 sample_data          | 2023-10-23T13:33:34.937Z  |  172.21.0.5    |  1232382              |  Disconnected
 sample_data          | 2023-10-23T12:27:28.948Z  |  172.21.2.113  |  2764889              |  Connected to 10.1.0.2
 sample_data          | 2023-10-23T12:15:03.360Z  |  172.21.2.162  |  3450233              |  Connected to 10.1.0.3
+sample_data_mv       | 2023-10-23T13:55:01.543Z  |  172.21.3.15   |  1756467              |  [Banana, Connected to 10.1.0.1]
```
Just renamed to mv_sample_data and discarded the changes on this file

ivancea committed:
Fixed failing test after indices change

ivancea committed:
Merge branch 'main' into esql_blockh2

ivancea committed:
Unmute Categorize tests

ivancea committed:
Added a bunch of other tests from PR suggestions

ivancea commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/TestCaseSupplier.java:
```diff
@@ -1431,6 +1431,34 @@ public static TestCase typeError(List<TypedData> data, String expectedTypeError)
+                foldingExceptionClass,
+                foldingExceptionMessage,
+                extra,
+                data.stream().allMatch(d -> d.forceLiteral || DataType.isRepresentable(d.type))
```
Uhmm, after running some hundreds of thousands of tests (functions), looks like it was just ignoring some "counter", "period" and "duration" cases from 4 functions.

In any case, this piece was moved from existing code. I'll work on this on a different issue/PR, as it's unrelated. But good catch, I wonder if we're magically ignoring tests that didn't make sense to begin with

alex-spies replied:
I wasn't aware this was moved. If that's the case, no need to follow up on this. Thanks for looking into it!

ivancea committed:
Removed semantic text index, which is failing in cross cluster

ivancea committed:
Improved optimizer tests on categorize

ivancea commented on muted-tests.yml:
```diff
@@ -229,9 +214,9 @@ tests:
-- class: org.elasticsearch.xpack.esql.qa.single_node.FieldExtractorIT
-  method: testConstantKeywordField
-  issue: https://github.com/elastic/elasticsearch/issues/117524
+- class: org.elasticsearch.xpack.esql.qa.single_node.EsqlSpecIT
+  method: test {categorize.Mv via eval const *}
+  issue: https://github.com/elastic/elasticsearch/issues/114714
```
Actually, I just removed those, as the fix is in main already. But I probably missed this one. Unmuting it too

ivancea committed:
Unmute last categorize test

alex-spies reviewed (approved):
Heya, let's get this green and into main. The only work before merging is, IMO, getting the mutes/capabilities right and having correct expectations for the csv test with nulls - even if that means it needs muting, see below.

There is some immediate follow-up work that needs to be done, but that can be in subsequent PRs. I'm summarizing it here because the many comments are hard to navigate.

Remaining csv test cases:
- `STATS a = c, b = c BY c = CATEGORIZE(message)`
- `from test | STATS MV_COUNT(cat), COUNT(*) BY cat = CATEGORIZE(first_name)`
- `| stats mv_count(categorize(message)) by categorize(message)`

Correct hashing of multivalues (+ test against regressions), see https://github.com/elastic/elasticsearch/pull/114317#discussion_r1856908043.

Block hash tests:
- more cases https://github.com/elastic/elasticsearch/pull/114317#discussion_r1856883763
- stronger assertion https://github.com/elastic/elasticsearch/pull/114317#discussion_r1836996970

FoldNull: check if this change is necessary and add a comment as to why if that's still the case https://github.com/elastic/elasticsearch/pull/114317#discussion_r1856475770

Ideally:
- simplify the changes to CombineProjections https://github.com/elastic/elasticsearch/pull/114317#discussion_r1856524724

ivancea committed:
Rename capability to CATEGORIZE_V2

ivancea committed:
Ignore and fix null tests

ivancea commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/AbstractCategorizeBlockHash.java:
```diff
@@ -0,0 +1,105 @@
+ * Base BlockHash implementation for {@code Categorize} grouping function.
+ */
+public abstract class AbstractCategorizeBlockHash extends BlockHash {
+    // TODO: this should probably also take an emitBatchSize
```
@nik9000 Some info on this?

ivancea committed:
Fixed multivalues on Categorize and added tests using other fields

ivancea committed:
Use checked circuit breaker and fix CategorizeBlockHashTests

ivancea committed:
Restore Aggregate computeReferences()

ivancea committed:
Remove AggregatorMode parameter from BlockHash.build(), and format all

jan-elastic commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/AbstractCategorizeBlockHash.java:
```diff
@@ -0,0 +1,105 @@
+        }
+        try (BytesStreamOutput out = new BytesStreamOutput()) {
+            // TODO be more careful here.
+            out.writeVInt(categorizer.getCategoryCount());
+            for (SerializableTokenListCategory category : categorizer.toCategoriesById()) {
+                category.writeTo(out);
+            }
+            // We're returning a block with N positions just because the Page must have all blocks with the same position count!
+            return blockFactory.newConstantBytesRefBlockWith(out.bytes().toBytesRef(), categorizer.getCategoryCount());
```
Not sure what you exactly mean. The number of categories is not equal to the number of inputs texts, meaning you still have a mismatch in number of positions.

ivancea replied:
We're building here the intermediate state to pass to the `CategorizeIntermediateHashBlock`, with 1 row/position per category. So I imagine we can do this in 2 ways:

The current one:
```
Serialize into a BytesRef with:
- int (# of categories)
- every category

And send this BytesRef in a block with N "simulated" rows
```

Instead, do:
```
Write a block with a category (Serialized in a BytesRef) per position/row.
So we don't simulate anything, and we could even consume just 2 categories later and discard the rest (Maybe? Just a _funny_ possibility, not sure if it's possible)
```

nik9000 replied:
My memory was that the state was one blob of bytes and not a blob per category. There's, like, shared state. But it's been a month since I thought a lot about this. And I'm wrong about lots of stuff.

ivancea committed:
Merge branch 'main' into esql_blockh2

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  8.x  | Commit could not be cherrypicked due to conflicts |

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 114317`

nik9000 commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/AbstractCategorizeBlockHash.java:
```diff
@@ -0,0 +1,105 @@
+ * Base BlockHash implementation for {@code Categorize} grouping function.
+ */
+public abstract class AbstractCategorizeBlockHash extends BlockHash {
+    // TODO: this should probably also take an emitBatchSize
```
TLDR: It's probably not important for the single-element `BlockHash` implementations like this.

So `emitBatchSize` is a request to call `AddInput#add` every `emitBatchSize` entries. It's designed to prevent building a huge page of ordinals when processing `STATS BY a, b` and `a` and `b` are no single valued - especially if they are *both* multivaued. There it's the contract to put the row into an ordering for all combinations of `a` and `b` values. Since that can explode into a huge number of rows, we batch it.

This is much much less important for single element `BlockHash` implementations. They don't change the number of output rows. That's true even for CATEGORIZE. And if the incoming page already "wasn't too big" then the page of ordinals passed to the aggs can't be *that* big either.

When I first built this I thought I might apply this to single valued `BlockHash` implementations as well. It'd be consistent. It's lame to ignore this request. But it isn't *important* so I never got to it.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/FoldNull.java:
```diff
@@ -42,6 +43,7 @@ public Expression rule(Expression e) {
             }
         } else if (e instanceof Alias == false
             && e.nullable() == Nullability.TRUE
+            && e instanceof Categorize == false
```
More follow-up I think.

nik9000 commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:
```diff
@@ -1203,6 +1204,26 @@ public void testCombineProjectionWithAggregationFirstAndAliasedGroupingUsedInAgg
+        var plan = plan("""
+            from test
+            | eval k = first_name, k1 = k
+            | stats s = sum(salary) by cat = CATEGORIZE(k)
```
More for the followup?

alex-spies replied:
I'm adding a reminder via an assertion in https://github.com/elastic/elasticsearch/pull/117882.
